core/src/main/java/org/apache/calcite/sql/fun/SqlTranslateFunction.java

```diff
@@ -78,6 +85,40 @@ protected SqlTranslateFunction(String name) {
     super.validateQuantifier(validator, call);
   }
 
+  @Override public RelDataType inferReturnType(
+      SqlOperatorBinding opBinding) {
+    final RelDataTypeFactory typeFactory = opBinding.getTypeFactory();
+    final RelDataType ret = opBinding.getOperandType(0);
+    if (SqlTypeUtil.isNull(ret)) {
+      return ret;
+    }
+    final String descCharsetName;
+    if (opBinding instanceof SqlCallBinding) {
+      descCharsetName = ((SqlCallBinding) opBinding).getCall().operand(1).toString();
+    } else {
+      descCharsetName = ((RexCallBinding) opBinding).getStringLiteralOperand(1);
+    }
+    assert descCharsetName != null;
+    Charset descCharset = SqlUtil.getCharset(descCharsetName);
+    return typeFactory.createTypeWithCharsetAndCollation(ret, descCharset, getCollation(ret));
+  }
+
+  @Override public RelDataType deriveType(SqlValidator validator,
+      SqlValidatorScope scope, SqlCall call) {
+    // special case for TRANSLATE: don't need to derive type for Charsets
```

Comment:
it's not obvious what this comment refers to. Special compared to what?

Reply:
Got your point, I'll change it to `don't need to derive type for Charsets` in both CONVERT and TRANSLATE. WDYT?